service/history/shard/context.go

```diff
@@ -624,67 +619,55 @@ func (s *contextImpl) CreateWorkflowExecution(
 		return nil, err
 	}
 
-Create_Loop:
-	for attempt := 0; attempt < conditionalRetryCount && ctx.Err() == nil; attempt++ {
-		if s.isClosed() {
-			return nil, ErrShardClosed
-		}
-		currentRangeID := s.getRangeID()
-		request.RangeID = currentRangeID
-
-		response, err := s.executionManager.CreateWorkflowExecution(ctx, request)
-		switch err.(type) {
-		case nil:
-			// Update MaxReadLevel if write to DB succeeds
-			s.updateMaxReadLevelLocked(transferMaxReadLevel)
-			return response, nil
-		case *types.WorkflowExecutionAlreadyStartedError,
-			*persistence.WorkflowExecutionAlreadyStartedError,
-			*persistence.CurrentWorkflowConditionFailedError,
-			*types.ServiceBusyError:
-			// No special handling required for these errors
-			// We know write to DB fails if these errors are returned
+	if s.isClosed() {
+		return nil, ErrShardClosed
+	}
+	currentRangeID := s.getRangeID()
+	request.RangeID = currentRangeID
+
+	response, err := s.executionManager.CreateWorkflowExecution(ctx, request)
+	switch err.(type) {
+	case nil:
+		// Update MaxReadLevel if write to DB succeeds
+		s.updateMaxReadLevelLocked(transferMaxReadLevel)
+		return response, nil
+	case *types.WorkflowExecutionAlreadyStartedError,
+		*persistence.WorkflowExecutionAlreadyStartedError,
+		*persistence.CurrentWorkflowConditionFailedError,
+		*types.ServiceBusyError:
+		// No special handling required for these errors
+		// We know write to DB fails if these errors are returned
+		return nil, err
+	case *persistence.ShardOwnershipLostError:
+		{
+			// Shard is stolen, trigger shutdown of history engine
+			s.logger.Warn(
+				"Closing shard: CreateWorkflowExecution failed due to stolen shard.",
+				tag.Error(err),
+			)
+			s.closeShard()
 			return nil, err
-		case *persistence.ShardOwnershipLostError:
-			{
-				// RangeID might have been renewed by the same host while this update was in flight
-				// Retry the operation if we still have the shard ownership
-				if currentRangeID != s.getRangeID() {
-					continue Create_Loop
```

Comment:
This path can never be hit, because the rangeID is protected by a lock, and the lock is held during the write operation

Reply:
```
// RangeID might have been renewed by the same host while this update was in flight
// Retry the operation if we still have the shard ownership
```
hmmmmmmmmm.  this seems to imply an async update in this process, and `s.rangeID` is an atomic var...
... which is never read from.  only written to.

I agree that this only changes in renewRangeLocked and:
- before this should never be hit because it set `currentRangeID := s.getRangeID()` before entering this switch
- now this can never be hit because it doesn't loop at all, so it releases and re-acquires the lock before re-reading (no worse than before)

but this does make me wonder if the whole `contextImpl` may be flawed / using the wrong field or something.  or that atomic var is just another piece of dead code perhaps.

Reply:
ok, so rangeID was probably effectively removed in #2634